src/journalrw.rs

```diff
@@ -0,0 +1,545 @@
+    let user_id = parts_iter.next().and_then(|u| if u.is_empty() { None } else { Some(u.to_string()) });
+
+    Ok(JournalEntry {
+        epoch_msec,
+        path,
+        signal: domain.unwrap_or(SIG_CHNG).into(),
+        source: METH_GET.into(),
+        value,
+        access_level: AccessLevel::Read as _,
+        short_time,
+        user_id,
+        repeat: value_flags & (1 << VALUE_FLAG_SPONTANEOUS_BIT) == 0,
+        provisional: value_flags & (1 << VALUE_FLAG_PROVISIONAL_BIT) != 0,
+    })
+}
+
+pub struct JournalReaderLog2<R> {
+    lines: Lines<R>,
+}
+
+impl<R> JournalReaderLog2<R>
+where
+    R: AsyncBufRead + Unpin,
+{
+    pub fn new(reader: R) -> Self {
+        JournalReaderLog2 {
+            lines: reader.lines(),
+        }
+    }
+}
+
+impl<R> Stream for JournalReaderLog2<R>
+where
+    R: AsyncBufRead + Unpin,
+{
+    type Item = Result<JournalEntry, Box<dyn Error>>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let lines = Pin::new(&mut self.lines);
+
+        match lines.poll_next(cx) {
+            Poll::Ready(Some(Ok(line))) => Poll::Ready(Some(parse_journal_entry_log2(&line))),
+            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(Box::new(e)))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+pub struct JournalWriterLog2<W> {
+    writer: W,
+}
+
+impl<W> JournalWriterLog2<W>
+where
+    W: AsyncWrite + Unpin,
+{
+    pub fn new(writer: W) -> Self {
+        Self {
+            writer,
+        }
+    }
+
+    pub async fn append(&mut self, entry: &JournalEntry)  -> Result<(), std::io::Error> {
+        let mut msec = entry.epoch_msec;
+        if msec == 0 {
+            msec = shvproto::DateTime::now().epoch_msec();
+        }
+        self.append_with_time(msec, msec, entry).await
+    }
+
+    pub async fn append_with_time(&mut self, msec: i64, orig_time: i64, entry: &JournalEntry) -> Result<(), std::io::Error> {
+        let line = {
+            let mut fields = vec![];
+            fields.push(shvproto::DateTime::from_epoch_msec(msec).to_iso_string());
+            fields.push(if orig_time == msec { "".into() } else { shvproto::DateTime::from_epoch_msec(orig_time).to_iso_string() });
+            fields.push(entry.path.clone());
+            fields.push(entry.value.to_cpon());
+            fields.push(if entry.short_time >= 0 { entry.short_time.to_string() } else { "".into() });
+            fields.push(entry.signal.clone());
+            let mut value_flags = 0u32;
+            if !entry.repeat {
+                value_flags |= 1 << VALUE_FLAG_SPONTANEOUS_BIT;
+            }
+            if entry.provisional {
+                value_flags |= 1 << VALUE_FLAG_PROVISIONAL_BIT;
+            }
+            fields.push(value_flags.to_string());
+            fields.push(entry.user_id.clone().unwrap_or_default());
+            fields.join(JOURNAL_ENTRIES_SEPARATOR) + "\n"
```

Comment:
I was thinking something like this:
```suggestion
            let mut value_flags = 0u32;
            if !entry.repeat {
                value_flags |= 1 << VALUE_FLAG_SPONTANEOUS_BIT;
            }
            if entry.provisional {
                value_flags |= 1 << VALUE_FLAG_PROVISIONAL_BIT;
            }

            [
                shvproto::DateTime::from_epoch_msec(msec).to_iso_string(),
                if orig_time == msec { "".into() } else { shvproto::DateTime::from_epoch_msec(orig_time).to_iso_string() },
                entry.path.clone(),
                entry.value.to_cpon(),
                if entry.short_time >= 0 { entry.short_time.to_string() } else { "".into() },
                entry.signal.clone(),
                value_flags.to_string(),
                entry.user_id.clone().unwrap_or_default(),
            ].join(JOURNAL_ENTRIES_SEPARATOR) + "\n"
```